### PR TITLE
Add new TargetBuiltReasons

### PR DESCRIPTION
**Pull request by maridematte**
Fixes https://github.com/dotnet/msbuild/issues/9467

### Context
When building targets we only specify a reason when it has direct control over the flow. So some targets have no built reason at the end of the build and can get very confusing on large builds.

### Changes Made
Added options for `TargetBuiltReason`: Initial target, entry target, default target.
These are added to `TargetSpecification` so they can be referenced in different points of the build.

### CODE REVIEW

**Review by JanKrivanek — COMMENTED** [review]
Thank you for diving into this!

It looks as a good direction. I believe we should test the case of explicit entry target (or implicit default target) and identical initial target - as then the same target should be executed twice - for 2 different reasons. The current implementation seem to not cover those cases (both executions would be attributed to initial target).
Other than that - I like the solution - simple and effective

**Comment by rainersigwald** [comment]
> I believe we should test the case of explicit entry target (or implicit default target) and identical initial target - as then the same target should be executed twice - for 2 different reasons.

Not sure I understand this. A target must be executed only once, for exactly one reason--even if it is put in the target stack multiple times for multiple reasons. Right?

**Comment by JanKrivanek** [comment]
> > I believe we should test the case of explicit entry target (or implicit default target) and identical initial target - as then the same target should be executed twice - for 2 different reasons.
> 
> Not sure I understand this. A target must be executed only once, for exactly one reason--even if it is put in the target stack multiple times for multiple reasons. Right?

I believe for InitialTargets we do not deduplicate this and we document it so as well - but will double check (afk now though). In general - target can execute multiple times if not properly input/output attributed, right? Or maybe I'm just confused :-)

**Comment by rainersigwald** [comment]
> In general - target can execute multiple times if not properly input/output attributed, right?

Within a single project instance in a single build, the engine should never run the target again, at most emitting `Target "{0}" skipped. Previously built successfully.` to the log.

**Comment by JanKrivanek** [comment]
Sorry - yes, you are correct - I was totally wrong on this.
I still have some recollection that even for skipped we could set the BuiltReason. But I can be wrong again. @maridematte - please ignore my comment until I check this on my end with code at hand

**Comment by JanKrivanek** [comment]
Seems like `TargetSkippedBuildEventArgs` has BuiltReason as well: https://github.com/dotnet/msbuild/blob/main/src/Framework/TargetSkippedEventArgs.cs#L102 It's being populated from the `TargetEntry` populated here. It's probably not used in the viewer - as viewer doesn't seem to display built reason for skipped targets (@KirillOsenkov ?). But even despite that it might not be bad idea to populate the information properly - we just need to pass a tupple list instead of a string list into the `BuildTargets`

**Comment by KirillOsenkov** [comment]
Reminds me of this case I think:
image

File: src/Framework/TargetBuiltReason.cs [hunk 1]
```diff
@@ -28,6 +28,22 @@ public enum TargetBuiltReason
...
         /// The target was part of the parent's AfterTargets list.
         /// </summary>
-        AfterTargets
+        AfterTargets,
+
+        /// <summary>
+        /// The target was defined as an initial target of the project.
+        /// </summary>
+        InitialTarget,
+
```
Comment (KirillOsenkov) [inline src/Framework/TargetBuiltReason.cs:37]: maybe remove the empty line. As a rule in C# there shouldn't be two consecutive empty lines.

File: src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs [hunk 2]
```diff
@@ -737,7 +751,15 @@ private async Task<bool> PushTargets(IList<TargetSpecification> targets, TargetE
 
                 // Add to the list of targets to push.  We don't actually put it on the stack here because we could run into a circular dependency
                 // during this loop, in which case the target stack would be out of whack.
-                TargetEntry newEntry = new TargetEntry(_requestEntry, this as ITargetBuilderCallback, targetSpecification, baseLookup, parentTargetEntry, buildReason, _componentHost, stopProcessingOnCompletion);
+                TargetEntry newEntry;
+                if (buildReason == TargetBuiltReason.None)
+                {
+                    newEntry = new TargetEntry(_requestEntry, this as ITargetBuilderCallback, targetSpecification, baseLookup, parentTargetEntry, targetSpecification._targetBuiltReason, _componentHost, stopProcessingOnCompletion);
```
Comment (KirillOsenkov) [inline src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs:757]: Would be nice to avoid having two lines of code both creating TargetEntry. I think it'll be better if you mutate the buildReason variable and pass it as an argument, so you can avoid duplicating lines 757 and 761.

**Comment by KirillOsenkov** [comment]
It is indeed pretty elegant and simpler than I thought it would be!

I filed a viewer issue to expose the TargetBuiltReason for skipped targets.

Would be nice to check if we can pass through this information for skipped targets.

Jan, not sure what you mean by passing a tuple instead of a string list, can you link to the source code?

**Comment by KirillOsenkov** [comment]
Also, be very careful here:

https://github.com/dotnet/msbuild/blob/f7f0b1924904ec7f2792c0a7805a4d0e91ac8955/src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs#L710

We should reason about this check very carefully, because now instead of None it could be something else. If the intent here was to check for "anything else but AfterTargets", the condition needs to be updated for the new enum values.

Need a lot of careful scrutiny here, because we might accidentally introduce a very bad regression.

**Comment by KirillOsenkov** [comment]
`PushTargets` seems like a very non-trivial algorithm. I'd sleep better at night if it was covered by unit-tests. Wondering if we already have tests covering this method? 

We need to trace the control flow in this method and carefully determine which branch of the if/else is (or should be) taken for each of the new enum values.

File: src/Framework/TargetBuiltReason.cs [hunk 3]
```diff
@@ -28,6 +28,22 @@ public enum TargetBuiltReason
...
+
+        /// <summary>
+        /// The target was the default target of the project
+        /// </summary>
+        DefaultTarget,
```
Comment (danmoseley) [inline src/Framework/TargetBuiltReason.cs:42]: Pluralization should match the xml attribute app DefaultTargets and InitialTargets like AfterTargets.
Also EntryTargets adds that can be several (/t:a;b)

Comment (KirillOsenkov) [reply src/Framework/TargetBuiltReason.cs:42]: so, to clarify, all three new enum fields should be plural, right?

Comment (danmoseley) [reply src/Framework/TargetBuiltReason.cs:42]: Yes

**Comment by JanKrivanek** [comment]
> Jan, not sure what you mean by passing a tuple instead of a string list, can you link to the source code?

**tl;dr;:** Setting the reason should be responsibility of the code that requested the build (and hence knows the reason)


Basically the `BuildTargets` is an internal contract and called from a single place in code (not considering the unit tests):

https://github.com/dotnet/msbuild/blob/f25414599dcb07202a9f8bfd8125d04a6e5c14be/src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs#L1210C21-L1210C31

So it should be very easy to transfer the responsibility of deciding about the build reason from the `BuildTargets` method, which actually doesn't have the knowledge (and the proposal now is just best guessing) to the caller (above) who is requesting the targets to be build and hence has the exact knowledge of the reason.
In such case the `BuildTargets` would accept e.g. `(string, BuiltReason)[] targets` instead of current `string[] targets` - which is easily accomodable on a caller side, removes the uncertainity on the callee side and is not breaking any contracts.

(but we already agreed on this offline with @maridematte)

File: src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs [hunk 4]
```diff
@@ -737,7 +737,15 @@ private async Task<bool> PushTargets(IList<TargetSpecification> targets, TargetE
...
-                TargetEntry newEntry = new TargetEntry(_requestEntry, this as ITargetBuilderCallback, targetSpecification, baseLookup, parentTargetEntry, buildReason, _componentHost, _projectLoggingContext, stopProcessingOnCompletion);
+                TargetEntry newEntry;
+                if (buildReason == TargetBuiltReason.None)
+                {
+                    newEntry = new TargetEntry(_requestEntry, this as ITargetBuilderCallback, targetSpecification, baseLookup, parentTargetEntry, targetSpecification._targetBuiltReason, _componentHost, _projectLoggingContext, stopProcessingOnCompletion);
+                }
+                else
+                {
+                    newEntry = new TargetEntry(_requestEntry, this as ITargetBuilderCallback, targetSpecification, baseLookup, parentTargetEntry, buildReason, _componentHost, _projectLoggingContext, stopProcessingOnCompletion);
+                }
```
Comment (JanKrivanek) [inline src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs:748]: If only the build reson differs here  let's replace this with a single call where only the build reason is conditioned (e.g. via ternary operator).

This way it'll be easier to read.

**Review by JanKrivanek — APPROVED** [review]
Thank you!

I have one suggestion to consider

**Review by YuliiaKovalova — APPROVED** [review]
Looks good to me!

File: src/Build.UnitTests/BackEnd/RequestBuilder_Tests.cs [hunk 5]
```diff
@@ -356,16 +356,16 @@ internal void SetNewBuildRequests(FullyQualifiedBuildRequest[] requests)
...
-        public Task<BuildResult> BuildTargets(ProjectLoggingContext loggingContext, BuildRequestEntry entry, IRequestBuilderCallback callback, string[] targets, Lookup baseLookup, CancellationToken cancellationToken)
+        public Task<BuildResult> BuildTargets(ProjectLoggingContext loggingContext, BuildRequestEntry entry, IRequestBuilderCallback callback, (string, TargetBuiltReason)[] targets, Lookup baseLookup, CancellationToken cancellationToken)
         {
             _requestBuilderCallback = callback;
 
             if (cancellationToken.WaitHandle.WaitOne(1500))
             {
                 BuildResult result = new BuildResult(entry.Request);
-                foreach (string target in targets)
+                foreach ((string, TargetBuiltReason) target in targets)
```
Comment (KirillOsenkov) [inline src/Build.UnitTests/BackEnd/RequestBuilder_Tests.cs:366]: you can do `(string name, TargetBuiltReason reason)` and then `target.reason` instead of `target.Item1`

File: src/Build.UnitTests/BackEnd/RequestBuilder_Tests.cs [hunk 6]
```diff
@@ -388,9 +388,9 @@ public Task<BuildResult> BuildTargets(ProjectLoggingContext loggingContext, Buil
                 if (cancellationToken.WaitHandle.WaitOne(1500))
                 {
                     BuildResult result = new BuildResult(entry.Request);
-                    foreach (string target in targets)
+                    foreach ((string, TargetBuiltReason) target in targets)
                     {
-                        result.AddResultsForTarget(target, BuildResultUtilities.GetEmptyFailingTargetResult());
+                        result.AddResultsForTarget(target.Item1, BuildResultUtilities.GetEmptyFailingTargetResult());
```
Comment (KirillOsenkov) [inline src/Build.UnitTests/BackEnd/RequestBuilder_Tests.cs:393]: consider using tuple variable names everywhere for consistency

**Comment by KirillOsenkov** [comment]
I'd say use named tuples everywhere and pick the same names for consistency, e.g. `(string targetName, TargetBuiltReason builtReason)`

Or maybe create a struct such as `internal record struct TargetInfo(string Name, TargetBuiltReason BuiltReason)` and use it everywhere instead (this way we can add more info to it in the future, such as ParentTarget, if needed).

**Comment by KirillOsenkov** [comment]
Otherwise I think it's good! Avoids changing any actual logic that might be a regression

**Review by KirillOsenkov — APPROVED** [review]

**Review by rainersigwald — APPROVED** [review]
There's enough repetition of the string `(string name, TargetBuiltReason reason)` in this PR that I personally would lean toward a named struct, but we can always switch to that if we need to add a field.

**Review by rainersigwald — APPROVED** [review]
